lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2554,8 +2554,15 @@ namespace {
         if (operationFn->getParameters()->size() != 1)
           return false;
 
-        if (operationFn->getParameters()->get(0)->getInterfaceType()->isEqual(
-                classDecl->getInterfaceType()))
+        auto paramType =
+            operationFn->getParameters()->get(0)->getInterfaceType();
+        if (auto optionalType = paramType->getOptionalObjectType()) {
+          paramType = optionalType;
+        }
+        auto paramDecl = paramType->getAnyNominal();
+        // The parameter of the retain/release function should be pointer to the
+        // same FRT.
+        if (paramDecl != classDecl)
```

Comment:
We should also allow retain/release operations to take a base class of `classDecl`, e.g.
```
struct RefCountedBase { ... } SWIFT_SHARED_REFERENCE(...);
inline void RCRetain(RefCountedBase* r);
inline void RCRelease(RefCountedBase* r);

struct MyObject : RefCountedBase { ... }; // uses RCRetain and RCRelease
```

I'm not fully confident, but I think the compiler accepted this code before this change, so it would be a regression if we ban this pattern now. Could you please add a test for this?

Reply:
Hey Egor, thanks for the excellent example!
This was supported before this patch and would not be supported at the current state of the patch. But I want to mention that this was supported by mistake. Earlier check:

```
if (operationFn->getParameters()->get(0)->getInterfaceType()->isEqual(
                classDecl->getInterfaceType()))
          return false;
```
would not have rejected anything because even for cases when retain/release functions have pointers to the same FRT, this `isEqual` check would fail (also for int, bool, etc). It would fail for anything so we never `return false` from this statement. 

But yes, I'll have to correct my current patch to cover the case where in retain/release functions of derived FRT we can pointer to base FRT :)

Reply:
I tried to call 
`classDecl->isSuperclassOf(paramClassDecl)` after typecasting `paramDecl` using `dyn_cast<ClassDecl>(paramDecl)` into a new variable `paramClassDecl`. But this is returning `false` for our base/derived FRT cases. Is the base-derived relationship of C++ FRT not maintained when they are imported and converted into Swift's `ClassDecl` ?

Reply:
That's right, there is currently no inheritance relationship on the Swift side between these types. You would have to do this check with the C++ types, e.g. with `clang::Sema::IsDerivedFrom`.